iOS/Flows/Conversation/Cells/ConversationMessageCellLayout.swift

```diff
@@ -0,0 +1,120 @@
+//
+//  ConversationMessageCellLayout.swift
+//  Jibber
+//
+//  Created by Martin Young on 11/4/21.
+//  Copyright © 2021 Benjamin Dodgson. All rights reserved.
+//
+
+import Foundation
+
+class ConversationMessageCellLayout: UICollectionViewFlowLayout {
+
+    override class var layoutAttributesClass: AnyClass {
+        return ConversationMessageCellLayoutAttributes.self
+    }
+
+    override func layoutAttributesForElements(in rect: CGRect) -> [UICollectionViewLayoutAttributes]? {
+        guard let attributesInRect = super.layoutAttributesForElements(in: rect) else { return nil }
+
+        for attributes in attributesInRect {
+            guard let messageAttributes = attributes as? ConversationMessageCellLayoutAttributes else {
+                continue
+            }
+            self.update(attributes: messageAttributes)
+        }
+
+        return attributesInRect
+    }
+
+    override func layoutAttributesForItem(at indexPath: IndexPath) -> UICollectionViewLayoutAttributes? {
+        let attributes = super.layoutAttributesForItem(at: indexPath)
+
+        if let messageAttributes = attributes as? ConversationMessageCellLayoutAttributes {
+            self.update(attributes: messageAttributes)
+        }
+
+        return attributes
+    }
+
+    private func update(attributes: ConversationMessageCellLayoutAttributes) {
+        guard let collectionView = self.collectionView else { return }
+
+        let indexPath = attributes.indexPath
+        let totalItemsInSection = collectionView.numberOfItems(inSection: indexPath.section)
+
+        // The higher the cell's index, the closer it is to the front of the message stack.
+        let stackIndex = totalItemsInSection - indexPath.item - 1
+
+        // Objects closer to the front of the stack should be brighter.
+        let backgroundBrightness = 1 - CGFloat(stackIndex) * 0.05
+        var backgroundColor: UIColor = indexPath.section == 0 ? .lightGray : .gray
+        backgroundColor = backgroundColor.color(withBrightness: backgroundBrightness)
+
+        attributes.backgroundColor = backgroundColor
+
+        // Only show text for the front most item in each section.
+        attributes.shouldShowText = stackIndex == 0
+
+        if indexPath.section == 0 {
+            // The first section should have a bubble tail on its first item
+            attributes.shouldShowTail = stackIndex == totalItemsInSection - 1
+            attributes.bubbleTailOrientation = .up
+        }
+
+        if indexPath.section == 1 {
+            // The second section should have a tail on its last item
+            attributes.shouldShowTail = stackIndex == 0
+            attributes.bubbleTailOrientation = .down
+        }
+    }
```

Comment:
Figuring out the the color for each cell, the tail orientation and whether text should show.